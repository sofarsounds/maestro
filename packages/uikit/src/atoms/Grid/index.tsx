import styled, { css } from 'styled-components';

/**
 * TODO
 * - Auto-stretch columns to fit when no width is given
 *
 */

interface ColProps {
  xs?: number;
  sm?: number;
  md?: number;
  lg?: number;
}
export const Col = styled.div<ColProps>`
  ${({ theme, xs, sm, md, lg }) => css`
    grid-column-end: span ${xs || 1};

    ${sm &&
      theme.media.tablet`
      grid-column-end: span ${sm};
    `}

    ${md &&
      theme.media.laptop`
      grid-column-end: span ${md};
    `}

    ${lg &&
      theme.media.desktop`
      grid-column-end: span ${lg};
    `}
  `}
`;

// Grid
interface GridProps {
  cols?: number;
  gap?: number;
}
export default styled.div<GridProps>`
  ${({ cols, gap }) => css`
    display: grid;
    grid-template-columns: repeat(${cols || 12}, 1fr);
    grid-gap: ${gap || '20px'};
  `}
`;
