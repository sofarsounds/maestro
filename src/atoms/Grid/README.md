#Â Grid 
To implement a `Grid` into your project you'll need to add this import
```js
import { Grid, Col } from '@sofarsounds/maestro'
```

After adding the import you can use the Grid simply like this
```html
<Grid>
  <Col xs={12} sm={6}>
    A column
  </Col>
  <Col xs={12} sm={6}>
    A second column
  </Col>
</Grid>
```

## Grid Props

| Name          | Type         | Default         | Description                      |
| :------------ | :-----       | :-------------- | :------------------------------- |
| **children**  | `React.Node` |                 | The content displayed in the grid
| cols          | `number`     |                 | Optional prop if you want a grid that isn't based on 12 columns
| gap           | `string`     | 20px            | Optional prop if you need more gap between rows and columns
| colGap        | `string`     | 20px            | Optional prop if you want to customise the gap between columns
| rowGap        | `string`     | 20px            | Optional prop if you want to customise the gap between rows
| data-testid   | `string`     |                 | Optional prop for testing purposes

## Col Props

| Name          | Type         | Default         | Description                      |
| :------------ | :-----       | :-------------- | :------------------------------- |
| **children**  | `React.Node` |                 | The content displayed in the column
| xs            | `number`     | 12              | Number of columns taken up on `xs` devices
| sm            | `number`     |                 | Number of columns taken up on `sm` devices
| md            | `number`     |                 | Number of columns taken up on `md` devices
| lg            | `number`     |                 | Number of columns taken up on `lg` devices
| xl            | `number`     |                 | Number of columns taken up on `xl` devices
| data-testid   | `string`     |                 | Optional prop for testing purposes
