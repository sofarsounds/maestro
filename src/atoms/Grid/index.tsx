import styled, { css } from 'styled-components';

/**
 * TODO (but not relevant for now)
 * - Auto-stretch columns to fit when no width is given
 */

interface ColProps {
  xs?: number;
  sm?: number;
  md?: number;
  lg?: number;
  xl?: number;
  rowSpan?: number;
}
export const Col = styled.div<ColProps>`
  ${({ theme, xs, sm, md, lg, xl, rowSpan }) => css`
    grid-column-end: span ${xs || 12};
    grid-row: span ${rowSpan || 'auto'};

    ${sm &&
      theme.media.md`
      grid-column-end: span ${sm};
    `}

    ${md &&
      theme.media.lg`
      grid-column-end: span ${md};
    `}

    ${lg &&
      theme.media.lg`
      grid-column-end: span ${lg};
    `}

    ${xl &&
      theme.media.xl`
      grid-column-end: span ${xl};
    `}
  `}
`;

// Grid
interface GridProps {
  cols?: number;
  gap?: string;
  colGap?: string;
  rowGap?: string;
}
export default styled.div<GridProps>`
  ${({ cols, gap, colGap, rowGap, theme }) => css`
    width: 100%;
    display: grid;
    grid-template-columns: repeat(${cols || 12}, 1fr);
    grid-gap: ${gap || '15px'};

    ${theme.media.sm`
      grid-gap: ${gap || '30px'}
     `}

    ${colGap &&
      css`
        grid-column-gap: ${colGap};
      `};

    ${rowGap &&
      css`
        grid-row-gap: ${rowGap};
      `};
  `}
`;
