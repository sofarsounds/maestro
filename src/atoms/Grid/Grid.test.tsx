import React from 'react';
import { mountWithTheme } from '../../test';

import Grid, { Col } from './index';

describe('Grid', () => {
  it('renders the Grid correctly', () => {
    const grid = mountWithTheme(
      <Grid>
        <Col xs={12} sm={4} md={4} lg={4}>
          Col 1
        </Col>
        <Col xs={12} sm={4} md={4} lg={4}>
          Col 2
        </Col>
      </Grid>
    );

    expect(grid).toMatchSnapshot();
  });

  it('renders a Grid with a custom gap', () => {
    const grid = mountWithTheme(
      <Grid gap="100px">
        <Col>Col 1</Col>
        <Col>Col 2</Col>
      </Grid>
    );

    expect(grid).toMatchSnapshot();
  });

  it('renders a Grid with a custom column gap', () => {
    const grid = mountWithTheme(
      <Grid colGap="1px">
        <Col>Col 1</Col>
        <Col>Col 2</Col>
        <Col>Col 3</Col>
      </Grid>
    );

    expect(grid).toMatchSnapshot();
  });

  it('renders a Grid with a custom row gap', () => {
    const grid = mountWithTheme(
      <Grid colGap="1px">
        <Col xs={12}>Row 1</Col>
        <Col xs={12}>Row 2</Col>
        <Col xs={12}>Row 3</Col>
      </Grid>
    );

    expect(grid).toMatchSnapshot();
  });

  it('renders a Grid with a custom row and column gap', () => {
    const grid = mountWithTheme(
      <Grid colGap="1px" rowGap="10px">
        <Col xs={6}>Row 3</Col>
        <Col xs={6}>Row 1</Col>
        <Col xs={6}>Row 1</Col>
        <Col xs={6}>Row 2</Col>
        <Col xs={6}>Row 2</Col>
      </Grid>
    );

    expect(grid).toMatchSnapshot();
  });

  it('renders a Grid with a custom row height', () => {
    const grid = mountWithTheme(
      <Grid colGap="1px">
        <Col rows={2}>Col 1</Col>
        <Col>Col 2</Col>
      </Grid>
    );

    expect(grid).toMatchSnapshot();
  });
});
