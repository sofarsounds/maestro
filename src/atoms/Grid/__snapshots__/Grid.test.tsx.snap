// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders a Grid with a custom gap 1`] = `
.c1 {
  grid-column-end: span 12;
  grid-column-start: auto;
  grid-row: span auto;
}

.c0 {
  width: 100%;
  display: grid;
  grid-template-columns: repeat(12,1fr);
  grid-gap: 15px;
}

@media (min-width:576px) {
  .c0 {
    grid-gap: 100px;
  }
}

<Component
  theme={
    Object {
      "borderRadius": Object {
        "button": "4px",
        "default": "2px",
        "input": "2px",
        "large": "8px",
      },
      "colors": Object {
        "backToBlack": "#111111",
        "black": "#000000",
        "blackBetty": "#1E1E1E",
        "blueChristmas": "#3498DB",
        "blueMagic": "#34495E",
        "blueSmoke": "#949494",
        "brownSugar": "#9B525F",
        "checkBerry": "#98035A",
        "darkGrey": "#535353",
        "darkSmoke": "#767676",
        "green100": "#CFEACC",
        "green200": "#9DDCAC",
        "green300": "#72D08A",
        "green400": "#4FC571",
        "green50": "#E6F7EA",
        "green500": "#23BA57",
        "green600": "#10AD52",
        "green700": "#039842",
        "green800": "#008736",
        "green900": "#006721",
        "greenGarden": "#19BC9B",
        "greenOnions": "#8BC34B",
        "lightBlueSmoke": "#94949499",
        "macyGrey": "#DBDBDB",
        "mellowYellow": "#F1C410",
        "orangeCrush": "#F39C12",
        "paintItBlack": "#333333",
        "peaches": "#F99973",
        "pinkTriangle": "#E91E63",
        "primary": "#039842",
        "purpleHaze": "#3F51B5",
        "purplePills": "#9B59B6",
        "redRedWine": "#EF4423",
        "royalBlood": "#FF5000",
        "silverSprings": "#F5F5F5",
        "whiteDenim": "#FFFFFF",
        "yellowSubmarine": "#FFD426",
      },
      "dimensions": Object {
        "button": Object {
          "default": Object {
            "height": "44px",
            "padding": "0 25px",
          },
          "small": Object {
            "height": "40px",
            "padding": "0 15px",
          },
        },
        "buttonHeight": "44px",
        "containerWidth": Object {
          "default": "800px",
          "fullscreen": "100%",
          "large": "1000px",
          "small": "600px",
        },
        "inputHeight": "44px",
        "navbarHeight": Object {
          "lg": "84px",
          "md": "74px",
          "xs": "62px",
        },
        "radioBox": "20px",
      },
      "fontSizes": Object {
        "base": "16px",
        "body1": "16px",
        "body2": "14px",
        "button": "14px",
        "caption": "10px",
        "formGroupLabel": "14px",
        "h1": Object {
          "lg": "48px",
          "md": "34px",
          "xs": "24px",
        },
        "h2": Object {
          "lg": "34px",
          "md": "24px",
          "xs": "20px",
        },
        "h3": Object {
          "lg": "24px",
          "md": "20px",
          "xs": "16px",
        },
        "h4": Object {
          "lg": "20px",
          "md": "16px",
          "xs": "16px",
        },
        "h5": Object {
          "lg": "16px",
          "md": "14px",
          "xs": "14px",
        },
        "h6": Object {
          "lg": "14px",
          "md": "14px",
          "xs": "14px",
        },
        "overline": "10px",
        "title": Object {
          "lg": "54px",
          "md": "38px",
          "xs": "26px",
        },
      },
      "fonts": Object {
        "regular": "Open Sans",
      },
      "media": Object {
        "lg": [Function],
        "md": [Function],
        "sm": [Function],
        "xl": [Function],
        "xs": [Function],
      },
      "ruler": Array [
        0,
        4,
        8,
        12,
        16,
        20,
        24,
        28,
        32,
        36,
        40,
        44,
        48,
        52,
        56,
        60,
        64,
        68,
        72,
        76,
        80,
      ],
      "utils": Object {
        "transition": [Function],
      },
      "zIndex": Object {
        "navbar": 100,
      },
    }
  }
>
  <styled.div
    gap="100px"
  >
    <div
      className="c0"
    >
      <styled.div>
        <div
          className="c1"
        >
          Col 1
        </div>
      </styled.div>
      <styled.div>
        <div
          className="c1"
        >
          Col 2
        </div>
      </styled.div>
    </div>
  </styled.div>
</Component>
`;

exports[`Grid renders a grid with custom start points 1`] = `
.c1 {
  grid-column-end: span 6;
  grid-column-start: 4;
  grid-row: span auto;
}

.c2 {
  grid-column-end: span 6;
  grid-column-start: auto;
  grid-row: span auto;
}

.c0 {
  width: 100%;
  display: grid;
  grid-template-columns: repeat(12,1fr);
  grid-gap: 15px;
}

@media (min-width:576px) {
  .c1 {
    grid-column-start: 4;
    grid-column-end: span 6;
  }
}

@media (min-width:768px) {
  .c1 {
    grid-column-start: 4;
    grid-column-end: span 6;
  }
}

@media (min-width:992px) {
  .c1 {
    grid-column-start: 4;
    grid-column-end: span 6;
  }
}

@media (min-width:576px) {
  .c2 {
    grid-column-start: auto;
    grid-column-end: span 6;
  }
}

@media (min-width:768px) {
  .c2 {
    grid-column-start: auto;
    grid-column-end: span 6;
  }
}

@media (min-width:992px) {
  .c2 {
    grid-column-start: auto;
    grid-column-end: span 6;
  }
}

@media (min-width:576px) {
  .c0 {
    grid-gap: 30px;
  }
}

<Component
  theme={
    Object {
      "borderRadius": Object {
        "button": "4px",
        "default": "2px",
        "input": "2px",
        "large": "8px",
      },
      "colors": Object {
        "backToBlack": "#111111",
        "black": "#000000",
        "blackBetty": "#1E1E1E",
        "blueChristmas": "#3498DB",
        "blueMagic": "#34495E",
        "blueSmoke": "#949494",
        "brownSugar": "#9B525F",
        "checkBerry": "#98035A",
        "darkGrey": "#535353",
        "darkSmoke": "#767676",
        "green100": "#CFEACC",
        "green200": "#9DDCAC",
        "green300": "#72D08A",
        "green400": "#4FC571",
        "green50": "#E6F7EA",
        "green500": "#23BA57",
        "green600": "#10AD52",
        "green700": "#039842",
        "green800": "#008736",
        "green900": "#006721",
        "greenGarden": "#19BC9B",
        "greenOnions": "#8BC34B",
        "lightBlueSmoke": "#94949499",
        "macyGrey": "#DBDBDB",
        "mellowYellow": "#F1C410",
        "orangeCrush": "#F39C12",
        "paintItBlack": "#333333",
        "peaches": "#F99973",
        "pinkTriangle": "#E91E63",
        "primary": "#039842",
        "purpleHaze": "#3F51B5",
        "purplePills": "#9B59B6",
        "redRedWine": "#EF4423",
        "royalBlood": "#FF5000",
        "silverSprings": "#F5F5F5",
        "whiteDenim": "#FFFFFF",
        "yellowSubmarine": "#FFD426",
      },
      "dimensions": Object {
        "button": Object {
          "default": Object {
            "height": "44px",
            "padding": "0 25px",
          },
          "small": Object {
            "height": "40px",
            "padding": "0 15px",
          },
        },
        "buttonHeight": "44px",
        "containerWidth": Object {
          "default": "800px",
          "fullscreen": "100%",
          "large": "1000px",
          "small": "600px",
        },
        "inputHeight": "44px",
        "navbarHeight": Object {
          "lg": "84px",
          "md": "74px",
          "xs": "62px",
        },
        "radioBox": "20px",
      },
      "fontSizes": Object {
        "base": "16px",
        "body1": "16px",
        "body2": "14px",
        "button": "14px",
        "caption": "10px",
        "formGroupLabel": "14px",
        "h1": Object {
          "lg": "48px",
          "md": "34px",
          "xs": "24px",
        },
        "h2": Object {
          "lg": "34px",
          "md": "24px",
          "xs": "20px",
        },
        "h3": Object {
          "lg": "24px",
          "md": "20px",
          "xs": "16px",
        },
        "h4": Object {
          "lg": "20px",
          "md": "16px",
          "xs": "16px",
        },
        "h5": Object {
          "lg": "16px",
          "md": "14px",
          "xs": "14px",
        },
        "h6": Object {
          "lg": "14px",
          "md": "14px",
          "xs": "14px",
        },
        "overline": "10px",
        "title": Object {
          "lg": "54px",
          "md": "38px",
          "xs": "26px",
        },
      },
      "fonts": Object {
        "regular": "Open Sans",
      },
      "media": Object {
        "lg": [Function],
        "md": [Function],
        "sm": [Function],
        "xl": [Function],
        "xs": [Function],
      },
      "ruler": Array [
        0,
        4,
        8,
        12,
        16,
        20,
        24,
        28,
        32,
        36,
        40,
        44,
        48,
        52,
        56,
        60,
        64,
        68,
        72,
        76,
        80,
      ],
      "utils": Object {
        "transition": [Function],
      },
      "zIndex": Object {
        "navbar": 100,
      },
    }
  }
>
  <styled.div>
    <div
      className="c0"
    >
      <styled.div
        lg={6}
        lgStart={4}
        md={6}
        mdStart={4}
        sm={6}
        smStart={4}
        xs={6}
        xsStart={4}
      >
        <div
          className="c1"
        >
          Col 1 / Row 1
        </div>
      </styled.div>
      <styled.div
        lg={6}
        md={6}
        sm={6}
        xs={6}
      >
        <div
          className="c2"
        >
          Col 1 / Row 2
        </div>
      </styled.div>
    </div>
  </styled.div>
</Component>
`;

exports[`Grid renders the Grid correctly 1`] = `
.c1 {
  grid-column-end: span 12;
  grid-column-start: auto;
  grid-row: span auto;
}

.c0 {
  width: 100%;
  display: grid;
  grid-template-columns: repeat(12,1fr);
  grid-gap: 15px;
}

@media (min-width:576px) {
  .c1 {
    grid-column-start: auto;
    grid-column-end: span 4;
  }
}

@media (min-width:768px) {
  .c1 {
    grid-column-start: auto;
    grid-column-end: span 4;
  }
}

@media (min-width:992px) {
  .c1 {
    grid-column-start: auto;
    grid-column-end: span 4;
  }
}

@media (min-width:576px) {
  .c0 {
    grid-gap: 30px;
  }
}

<Component
  theme={
    Object {
      "borderRadius": Object {
        "button": "4px",
        "default": "2px",
        "input": "2px",
        "large": "8px",
      },
      "colors": Object {
        "backToBlack": "#111111",
        "black": "#000000",
        "blackBetty": "#1E1E1E",
        "blueChristmas": "#3498DB",
        "blueMagic": "#34495E",
        "blueSmoke": "#949494",
        "brownSugar": "#9B525F",
        "checkBerry": "#98035A",
        "darkGrey": "#535353",
        "darkSmoke": "#767676",
        "green100": "#CFEACC",
        "green200": "#9DDCAC",
        "green300": "#72D08A",
        "green400": "#4FC571",
        "green50": "#E6F7EA",
        "green500": "#23BA57",
        "green600": "#10AD52",
        "green700": "#039842",
        "green800": "#008736",
        "green900": "#006721",
        "greenGarden": "#19BC9B",
        "greenOnions": "#8BC34B",
        "lightBlueSmoke": "#94949499",
        "macyGrey": "#DBDBDB",
        "mellowYellow": "#F1C410",
        "orangeCrush": "#F39C12",
        "paintItBlack": "#333333",
        "peaches": "#F99973",
        "pinkTriangle": "#E91E63",
        "primary": "#039842",
        "purpleHaze": "#3F51B5",
        "purplePills": "#9B59B6",
        "redRedWine": "#EF4423",
        "royalBlood": "#FF5000",
        "silverSprings": "#F5F5F5",
        "whiteDenim": "#FFFFFF",
        "yellowSubmarine": "#FFD426",
      },
      "dimensions": Object {
        "button": Object {
          "default": Object {
            "height": "44px",
            "padding": "0 25px",
          },
          "small": Object {
            "height": "40px",
            "padding": "0 15px",
          },
        },
        "buttonHeight": "44px",
        "containerWidth": Object {
          "default": "800px",
          "fullscreen": "100%",
          "large": "1000px",
          "small": "600px",
        },
        "inputHeight": "44px",
        "navbarHeight": Object {
          "lg": "84px",
          "md": "74px",
          "xs": "62px",
        },
        "radioBox": "20px",
      },
      "fontSizes": Object {
        "base": "16px",
        "body1": "16px",
        "body2": "14px",
        "button": "14px",
        "caption": "10px",
        "formGroupLabel": "14px",
        "h1": Object {
          "lg": "48px",
          "md": "34px",
          "xs": "24px",
        },
        "h2": Object {
          "lg": "34px",
          "md": "24px",
          "xs": "20px",
        },
        "h3": Object {
          "lg": "24px",
          "md": "20px",
          "xs": "16px",
        },
        "h4": Object {
          "lg": "20px",
          "md": "16px",
          "xs": "16px",
        },
        "h5": Object {
          "lg": "16px",
          "md": "14px",
          "xs": "14px",
        },
        "h6": Object {
          "lg": "14px",
          "md": "14px",
          "xs": "14px",
        },
        "overline": "10px",
        "title": Object {
          "lg": "54px",
          "md": "38px",
          "xs": "26px",
        },
      },
      "fonts": Object {
        "regular": "Open Sans",
      },
      "media": Object {
        "lg": [Function],
        "md": [Function],
        "sm": [Function],
        "xl": [Function],
        "xs": [Function],
      },
      "ruler": Array [
        0,
        4,
        8,
        12,
        16,
        20,
        24,
        28,
        32,
        36,
        40,
        44,
        48,
        52,
        56,
        60,
        64,
        68,
        72,
        76,
        80,
      ],
      "utils": Object {
        "transition": [Function],
      },
      "zIndex": Object {
        "navbar": 100,
      },
    }
  }
>
  <styled.div>
    <div
      className="c0"
    >
      <styled.div
        lg={4}
        md={4}
        sm={4}
        xs={12}
      >
        <div
          className="c1"
        >
          Col 1
        </div>
      </styled.div>
      <styled.div
        lg={4}
        md={4}
        sm={4}
        xs={12}
      >
        <div
          className="c1"
        >
          Col 2
        </div>
      </styled.div>
    </div>
  </styled.div>
</Component>
`;
